islandora_multi_importer.api.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+/**
+ * @file
+ * This file documents all available hook functions to manipulate data.
+ */
+
+/**
+ * Implements hook_islandora_multi_importer_remote_file_get for https://digital.grinnell.edu.
+ *
+ * @param $url
+ *   Path to the file to be fetched.  Behavior of this hook depends on the transfer scheme specified.
+ *     Default behavior strips the 'basename' from $url and looks for that file somewhere in public://imi_files.
+ *     @TODO... if smb:// is prepended to the $url implement a Samba file fetch here.
+ * @return array
+ *   On error or failure an empty array is returned.
+ *   On success, a single array element LOCAL path like /var/www/drupal7/sites/default/files/imi_files/exmample.xml
+ */
+function hook_islandora_multi_importer_remote_file_get($url) {
```

Comment:
When I tested this, IMI could **NOT properly resolve a local reference to public://**, presumably because of the issue noted above.

My implementation of the hook has also changed since this was submitted, but that's the beauty of providing a hook...it lets the user/developer take control of their own process.  And there's one BIG difference between my hook and the standard local* behavior, my hook implementation always does the following...

  ```  
  $target = "/mnt/storage";
  $basename = drupal_basename($parsed_url['path']);
  $pattern = "/" . preg_quote($basename) . "/";
  $options['key'] = 'filename';
  $file = file_scan_directory($target, $pattern, $options);
  ```

It builds a regular expression containing just the filename portion of the specified path, and then recursively scans the $target directory structure looking for a match.  That allows the folks who prep data for IMI to put there files in any directory structure, ANYWHERE below the $target path, and IMI will still find them, or report that it cannot.  This is huge, especially for Oral History objects that contain TWO datastream files per object:  an MP3 OBJ datastream, and its corresponding TRANSCRIPT XML file.  It's customary to put each pair of files in its own subdirectory, and with my hook implementation users are free to do that without having to specify precise something like this...

```
public://Volumes/libstu/AlumniOralHistories/Complete/ReadyForIMI/2012_Ted_Amann_87/Ted_Amann_87.MP3
public://Volumes/libstu/AlumniOralHistories/Complete/ReadyForIMI/2012_Ted_Amann_87/IOH-Ted_Amann_87.XML
```
...instead they just enter...

```
Ted_Amann_87.MP3
IOH-Ted_Amann_87.XML
```

...and let the hook do the heavy lifting.   

That is the beauty of providing a hook.  If you don't define one, fine, you get the standard behavior with very little overhead.  But if you do define one, you get control of the process to fit your workflow.

Reply:
Ok, i totally get the beauty of the hook. I also need to make sure public:// works too, but that is my responsibility and i will fix that. I will need probably a test suite here. A call at 15:00 today? Does that work for you? Send me an invite of some link to a conference app fo your liking.

Reply:
Sounds good.  I'll assume you mean 3 PM Eastern time, right?  I'll send an invite shortly.

Also, I'm going to include in this PR my latest hook implementation as islandora_multi_importer.api.php...  It's a much better implementation than I had when I was testing and trying to figure things out.

Reply:
Mark, can you update the hook example to your new implementation? Looks simpler and cleaner?

Reply:
Just updated the PR to include my latest hook implementation as documentation in islandora_multi_importer.api.php.